superbench/benchmarks/micro_benchmarks/directx_third_party/DeviceResources.cpp

```diff
@@ -0,0 +1,670 @@
+    }
+}
+
+// Destructor for DeviceResources.
+DeviceResources::~DeviceResources() {
+    // Ensure that the GPU is no longer referencing resources that are about to be destroyed.
+    WaitForGpu();
+}
+
+// Configures the Direct3D device, and stores handles to it and the device context.
+void DeviceResources::CreateDeviceResources() {
+#if defined(_DEBUG)
+    // Enable the debug layer (requires the Graphics Tools "optional feature").
+    //
+    // NOTE: Enabling the debug layer after device creation will invalidate the active device.
+    {
+        ComPtr<ID3D12Debug> debugController;
+        if (SUCCEEDED(D3D12GetDebugInterface(IID_PPV_ARGS(debugController.GetAddressOf())))) {
+            debugController->EnableDebugLayer();
+        } else {
+            OutputDebugStringA("WARNING: Direct3D Debug Device is not available\n");
+        }
+
+        ComPtr<IDXGIInfoQueue> dxgiInfoQueue;
+        if (SUCCEEDED(DXGIGetDebugInterface1(0, IID_PPV_ARGS(dxgiInfoQueue.GetAddressOf())))) {
+            m_dxgiFactoryFlags = DXGI_CREATE_FACTORY_DEBUG;
+
+            dxgiInfoQueue->SetBreakOnSeverity(DXGI_DEBUG_ALL, DXGI_INFO_QUEUE_MESSAGE_SEVERITY_ERROR, true);
+            dxgiInfoQueue->SetBreakOnSeverity(DXGI_DEBUG_ALL, DXGI_INFO_QUEUE_MESSAGE_SEVERITY_CORRUPTION, true);
+
+            DXGI_INFO_QUEUE_MESSAGE_ID hide[] = {
+                80 /* IDXGISwapChain::GetContainingOutput: The swapchain's adapter does not control the output on which
+                      the swapchain's window resides. */
+                ,
+            };
+            DXGI_INFO_QUEUE_FILTER filter = {};
+            filter.DenyList.NumIDs = static_cast<UINT>(std::size(hide));
+            filter.DenyList.pIDList = hide;
+            dxgiInfoQueue->AddStorageFilterEntries(DXGI_DEBUG_DXGI, &filter);
+        }
+    }
+#endif
+
+    ThrowIfFailed(CreateDXGIFactory2(m_dxgiFactoryFlags, IID_PPV_ARGS(m_dxgiFactory.ReleaseAndGetAddressOf())));
+
+    // Determines whether tearing support is available for fullscreen borderless windows.
+    if (m_options & c_AllowTearing) {
+        BOOL allowTearing = FALSE;
+        HRESULT hr =
+            m_dxgiFactory->CheckFeatureSupport(DXGI_FEATURE_PRESENT_ALLOW_TEARING, &allowTearing, sizeof(allowTearing));
+        if (FAILED(hr) || !allowTearing) {
+            m_options &= ~c_AllowTearing;
+#ifdef _DEBUG
+            OutputDebugStringA("WARNING: Variable refresh rate displays not supported");
+#endif
+        }
+    }
+
+    ComPtr<IDXGIAdapter1> adapter;
+    GetAdapter(adapter.GetAddressOf());
+
+    // Create the DX12 API device object.
+    HRESULT hr =
+        D3D12CreateDevice(adapter.Get(), m_d3dMinFeatureLevel, IID_PPV_ARGS(m_d3dDevice.ReleaseAndGetAddressOf()));
+    ThrowIfFailed(hr);
+
+    m_d3dDevice->SetName(L"DeviceResources");
+
+#ifndef NDEBUG
+    // Configure debug device (if active).
+    ComPtr<ID3D12InfoQueue> d3dInfoQueue;
+    if (SUCCEEDED(m_d3dDevice.As(&d3dInfoQueue))) {
+#ifdef _DEBUG
```

Comment:
>_DEBUG [](http://example.com/codeflow?start=7&length=6)

why do we check NDEBUG and _DEBUG here? Can we unify them?

Reply:
I may not change it since it's third party code